gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java

```diff
@@ -278,18 +281,34 @@ protected String getTimestampDisplay(long timestamp) {
    * Method that deletes expired tokens based on the token timestamp.
    */
   protected void evictExpiredTokens() {
-    try {
-      for (final String token : getTokens()) {
-        if (isExpired(token)) {
+    for (final String token : getTokens()) {
+      try {
+        if (needsEviction(token)) {
           log.evictToken(getTokenDisplayText(token));
           removeToken(token);
         }
+      } catch (final Exception e) {
+        log.failedExpiredTokenEviction(getTokenDisplayText(token), e);
       }
-    } catch(final Exception e) {
-      log.failedExpiredTokenEviction(e);
     }
   }
 
+  /**
+   * Method that checks if an expired token is ready to be evicted
+   * by adding configured grace period to the expiry time.
+   * @param token
+   * @return
+   */
+  protected boolean needsEviction(final String token) {
+    boolean needsEviction;
+    needsEviction = isUnknown(token); // Check if the token exist
```

Comment:
If the token is unknown in this case, you're going to get an IllegalArgumentException when the subsequent removal is attempted because this statement says an unknown token -> needsEviction. I would argue that an unknown token means it has probably already been evicted.
The following call to getTokenExpiration will already invoked validateToken(), so the up-front isUnknownToken check here is unnecessary (IMO) and the conclusion is incorrect.